### PR TITLE
Fix type narrowing edge case, bump to 0.7.1

### PR DESCRIPTION
**Pull request by brentyi**
Fixes:
- https://github.com/nerfstudio-project/nerfstudio/issues/2862
- https://github.com/brentyi/tyro/issues/121

Needs tests!

### CODE REVIEW

**Comment by brentyi** [comment]
Well this is not right, but the general idea is there...

File: src/tyro/__init__.py [hunk 1]
```diff
@@ -11,3 +11,7 @@
 # Deprecated interface.
 if not TYPE_CHECKING:
     from ._deprecated import *  # noqa
+
+
+# TODO: this should be synchronized automatically with the pyproject.toml.
+__version__ = "0.7.1"
```
Comment (pwais) [inline src/tyro/__init__.py:17]: @brentyi  tankuuuuuu!!

i have only ever done version sync thru setup.py but allegedly this works for pyproject.toml thingies  https://stackoverflow.com/a/67097076

Comment (brentyi) [reply src/tyro/__init__.py:17]: Maybe something wrong with environment, but I just tried this and unfortunately it seemed flakey for me... kept returning `0.6.0` after I installed `0.7.1`.
